libbeat/processors/actions/alterFieldProcessor.go

```diff
@@ -0,0 +1,145 @@
+	AlterFullField bool
+
+	processorName string
+	alterFunc     mapstr.AlterFunc
+}
+
+// NewAlterFieldProcessor is an umbrella method for processing events based on provided fields. Such as converting event keys to uppercase/lowercase
+func NewAlterFieldProcessor(c *conf.C, processorName string, alterFunc mapstr.AlterFunc) (beat.Processor, error) {
+	config := struct {
+		Fields         []string `config:"fields"`
+		IgnoreMissing  bool     `config:"ignore_missing"`
+		FailOnError    bool     `config:"fail_on_error"`
+		AlterFullField bool     `config:"alter_full_field"`
+	}{
+		IgnoreMissing:  false,
+		FailOnError:    true,
+		AlterFullField: true,
+	}
+
+	if err := c.Unpack(&config); err != nil {
+		return nil, fmt.Errorf("failed to unpack the %s fields configuration: %w", processorName, err)
+	}
+
+	// Skip mandatory fields
+	for _, readOnly := range processors.MandatoryExportedFields {
+		for i, field := range config.Fields {
+			if field == readOnly {
+				config.Fields = append(config.Fields[:i], config.Fields[i+1:]...)
+			}
+		}
+	}
+	return &alterFieldProcessor{
+		Fields:         config.Fields,
+		IgnoreMissing:  config.IgnoreMissing,
+		FailOnError:    config.FailOnError,
+		processorName:  processorName,
+		AlterFullField: config.AlterFullField,
+		alterFunc:      alterFunc,
+	}, nil
+
+}
+
+func (a *alterFieldProcessor) String() string {
+	return fmt.Sprintf("%s fields=%+v", a.processorName, *a)
+}
+
+func (a *alterFieldProcessor) Run(event *beat.Event) (*beat.Event, error) {
+	var backup *beat.Event
+	if a.FailOnError {
+		backup = event.Clone()
+	}
+
+	for _, field := range a.Fields {
+		err := a.alter(event, field)
+		if err != nil {
+			if a.IgnoreMissing && errors.Is(err, mapstr.ErrKeyNotFound) {
+				continue
+			}
+			if a.FailOnError {
+				event = backup
+				_, _ = event.PutValue("error.message", err.Error())
+				return event, err
+			}
+		}
+	}
+
+	return event, nil
+}
+
+func (a *alterFieldProcessor) alter(event *beat.Event, field string) error {
+
+	// modify all segments of the key
+	if a.AlterFullField {
+		err := event.Fields.AlterPath(field, mapstr.CaseInsensitiveMode, a.alterFunc)
+		if err != nil {
+			return err
+		}
+	} else {
+		// modify only the last segment
+		segmentCount := strings.Count(field, ".") + 1
+		err := event.Fields.Traverse(field, mapstr.CaseInsensitiveMode, func(level mapstr.M, key string) error {
+			segmentCount--
+			if segmentCount == 0 {
+				val := level[key]
+				newKey, err := a.alterFunc(key)
+				if err != nil {
+					return fmt.Errorf("failed to apply a change to %q: %w", key, err)
+				}
+				if newKey == "" {
+					return fmt.Errorf("replacement key for %q cannot be empty", key)
+				}
+				if newKey != key {
+					_, exists := level[newKey]
+					if exists {
+						return fmt.Errorf("replacement key %q already exists: %w", newKey, mapstr.ErrKeyCollision)
+					}
+					delete(level, key)
+					level[newKey] = val
+				}
+
+				return nil
+			}
+			return nil
+		})
```

Comment:
I think it can be much simpler:

```suggestion
		segmentCount := strings.Count(field, ".")
		err := event.Fields.AlterPath(field, mapstr.CaseInsensitiveMode, func(key string) (string, error) {
			if segmentCount > 0 {
				segmentCount--
				return key, nil
			}
			return a.alterFunc(key)
		})
```